From 3d7ad273ea1ad10a48daa9e257c412978a4f56a5 Mon Sep 17 00:00:00 2001
From: Don Zickus <dzickus@redhat.com>
Date: Tue, 16 Sep 2014 17:03:32 -0400
Subject: [usb] wusbcore: read actual_length bytes isoc in segments

Message-id: <1410887463-20674-141-git-send-email-dzickus@redhat.com>
Patchwork-id: 95162
O-Subject: [RHEL7 PATCH 140/591] usb: wusbcore: read actual_length bytes isoc in segments
Bugzilla: 1110939
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1110939

commit ecf3701cede840476b012ac4796f77c6dd9ee623
Author: Thomas Pugliese <thomas.pugliese@gmail.com>
Date:   Fri Feb 28 15:10:26 2014 -0600

    usb: wusbcore: read actual_length bytes isoc in segments

    Use the iso_frame_desc.actual_length field instead of length when
    reading isoc in data segments from the HWA.  This fixes a case where the
    isoc in read URB would never complete because it expected the HWA to
    send more data than it actually did.  When this happened the URB would
    be stuck in the driver preventing module unload and clean shutdown.

    Signed-off-by: Thomas Pugliese <thomas.pugliese@gmail.com>
    Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/usb/wusbcore/wa-xfer.c b/drivers/usb/wusbcore/wa-xfer.c
index 1b2fb75..4f21115 100644
--- a/drivers/usb/wusbcore/wa-xfer.c
+++ b/drivers/usb/wusbcore/wa-xfer.c
@@ -2160,7 +2160,7 @@ static void __wa_populate_buf_in_urb_isoc(struct wahc *wa, struct wa_xfer *xfer,
  wa->buf_in_urb->transfer_dma = xfer->urb->transfer_dma +
   xfer->urb->iso_frame_desc[curr_iso_frame].offset;
  wa->buf_in_urb->transfer_buffer_length =
-  xfer->urb->iso_frame_desc[curr_iso_frame].length;
+  xfer->urb->iso_frame_desc[curr_iso_frame].actual_length;
  wa->buf_in_urb->transfer_flags |= URB_NO_TRANSFER_DMA_MAP;
  wa->buf_in_urb->transfer_buffer = NULL;
  wa->buf_in_urb->sg = NULL;
-- 
1.7.1